src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,172 @@
+        tag_hash = Hash.new
+        vm_config_tags.each do |vm_config_tag|
+
+          if vm_config_tag["category"].nil?
+            logger.warn( "Missing category content, skip processing this category-tag pair." )
+            next
+          end
+          if vm_config_tag["category"].empty?
+            logger.warn( "Empty category found in cloud config ,skip processing this category-tag pair." )
+            next
+          end
+          if vm_config_tag["tag"].nil?
+            logger.warn( "Missing tag in category '#{vm_config_tag["category"]}', skip attaching this tag." )
+            next
+          end
+          if vm_config_tag["tag"].empty?
+            logger.warn( "Empty tag found in category '#{vm_config_tag["category"]}', skip attaching this tag." )
+            next
+          end
+          tag_hash[vm_config_tag["category"]] ||= []
+          tag_hash[vm_config_tag["category"]] << vm_config_tag["tag"]
+        end
+        tag_hash
+      end
+
+      def attach_single_tag(vm_mob_id, tag_id )
+        tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttach.new
+        tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        tag_assoc_info.object_id.id = vm_mob_id
+        tag_assoc_info.object_id.type = 'VirtualMachine'
+        tag_assoc_api.attach(tag_id,tag_assoc_info)
+      end
+
+      def attach_multi_tags(vm_mob_id, tag_ids, category_name, target_category_id)
+        single_cardinality = single_category_cardinality?(target_category_id)
+        if single_cardinality
+          logger.warn( "Tag category '#{category_name}' does not support 'Multiple Cardinality', skip tagging this category to vm." )
+          return
+        end
+        multi_tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttachMultipleTagsToObject.new
+        multi_tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        multi_tag_assoc_info.object_id.id = vm_mob_id
+        multi_tag_assoc_info.object_id.type = 'VirtualMachine'
+        multi_tag_assoc_info.tag_ids = tag_ids
+        tag_assoc_api.attach_multiple_tags_to_object(multi_tag_assoc_info)
+      end
+
+      def attach_tags(vm_mob_id, vm_config_tags, vm_config_name)
+        tag_hash = create_tag_hash(vm_config_tags)
+        if tag_hash.empty?
+          logger.warn( "No valid category-tag pair in cloud config, finish attaching tags." )
+          return
+        end
+        category_ids = tagging_category_api.list.value
+        if category_ids.empty?
+          logger.warn( "No category exist on vCenter, skip attaching tags." )
+          return
+        end
+
+        tag_hash.keys.each do |category_name|
+          begin
+            target_category_id = retrieve_category_id(category_name, category_ids)
+            if target_category_id.nil?
+              logger.warn( "Tag category '#{category_name}' does not exit in vCenter, skip attaching this category to vm." )
```

Comment:
I don't love this pattern. Instead we should raise an exception, catch it in a `rescue`, and do the `logger.warn` and `next` there.

Reply:
One benefit is that `attach_single_tag` and `attach_multi_tag` can raise the same type of exception to skip this tag.